homeassistant/helpers/state.py

```diff
@@ -145,76 +205,21 @@ def reproduce_state(hass: HomeAssistantType,
-    """Reproduce given state."""
+    """Reproduce a list of states on multiple domains."""
     if isinstance(states, State):
         states = [states]
 
-    to_call = defaultdict(list)  # type: Dict[Tuple[str, str, str], List[str]]
+    to_call = defaultdict(list)
 
     for state in states:
-
-        if hass.states.get(state.entity_id) is None:
-            _LOGGER.warning("reproduce_state: Unable to find entity %s",
-                            state.entity_id)
-            continue
-
-        if state.domain == GROUP_DOMAIN:
-            service_domain = HASS_DOMAIN
-        else:
-            service_domain = state.domain
-
-        domain_services = hass.services.async_services().get(service_domain)
-
-        if not domain_services:
-            _LOGGER.warning(
-                "reproduce_state: Unable to reproduce state %s (1)", state)
-            continue
-
-        service = None
-        for _service in domain_services.keys():
-            if (_service in SERVICE_ATTRIBUTES and
-                    all(attr in state.attributes
-                        for attr in SERVICE_ATTRIBUTES[_service]) or
-                    _service in SERVICE_TO_STATE and
-                    SERVICE_TO_STATE[_service] == state.state):
-                service = _service
-            if (_service in SERVICE_TO_STATE and
-                    SERVICE_TO_STATE[_service] == state.state):
-                break
-
-        if not service:
-            _LOGGER.warning(
-                "reproduce_state: Unable to reproduce state %s (2)", state)
-            continue
-
-        # We group service calls for entities by service call
-        # json used to create a hashable version of dict with maybe lists in it
-        key = (service_domain, service,
-               json.dumps(dict(state.attributes), sort_keys=True))
-        to_call[key].append(state.entity_id)
-
-    domain_tasks = {}  # type: Dict[str, List[Awaitable[Optional[bool]]]]
-    for (service_domain, service, service_data), entity_ids in to_call.items():
-        data = json.loads(service_data)
-        data[ATTR_ENTITY_ID] = entity_ids
-
-        if service_domain not in domain_tasks:
-            domain_tasks[service_domain] = []
-
-        domain_tasks[service_domain].append(
-            hass.services.async_call(service_domain, service, data, blocking)
-        )
-
-    async def async_handle_service_calls(
-            coro_list: Iterable[Awaitable]) -> None:
-        """Handle service calls by domain sequence."""
-        for coro in coro_list:
-            await coro
-
-    execute_tasks = [async_handle_service_calls(coro_list)
-                     for coro_list in domain_tasks.values()]
-    if execute_tasks:
-        await asyncio.wait(execute_tasks, loop=hass.loop)
+        to_call[state.domain].append(state)
+
+    if to_call:
+        await asyncio.gather(*[
+            getattr(hass.components, domain).async_reproduce_states(
```

Comment:
Do we need to handle the case of missing component for a domain?

Reply:
Yea. I was thinking of adding a fallback for the "standard" services turn_on/turn_off.